src/main/app/controller/courts/AddressController.ts

```diff
@@ -0,0 +1,276 @@
+          const errors = postcodeValidation.errors.length === 0
+            ? [this.updateAddressError] // we've encountered a 400 for a reason other than postcodes
+            : postcodeValidation.errors;
+          await this.render(req, res, false, addresses, errors, postcodeValidation.primaryInvalid, postcodeValidation.secondaryInvalid);
+        } else {
+          await this.render(req, res, false, addresses, [this.updateAddressError]);
+        }
+      });
+  }
+
+  private async render(
+    req: AuthedRequest,
+    res: Response,
+    updated = false,
+    addresses: DisplayCourtAddresses = null,
+    errorMsgs: string[] = [],
+    primaryPostcodeInvalid = false,
+    secondaryPostcodeInvalid = false) {
+
+    const slug: string = req.params.slug as string;
+    let fatalError = false;
+
+    if (!addresses) {
+      await req.scope.cradle.api.getCourtAddresses(slug)
+        .then((addressList: CourtAddress[]) => addresses = this.convertToDisplayAddresses(addressList))
+        .catch(() => {
+          errorMsgs.push(this.getAddressesError);
+          fatalError = true;
+        });
+    }
+
+    let addressTypes: AddressType[] = [];
+    await req.scope.cradle.api.getAddressTypes()
+      .then((types: AddressType[]) => addressTypes = types)
+      .catch(() => {
+        errorMsgs.push(this.getAddressTypesError);
+        fatalError = true;
+      });
+    const writeToUsTypes = addressTypes.filter(at => at.name.toLowerCase() === this.writeToUsAddressType.toLowerCase()).map(at => at.id);
+
+    const pageData: CourtAddressPageData = {
+      addressTypesPrimary: this.getAddressTypesForSelect(addressTypes, true),
+      addressTypesSecondary: this.getAddressTypesForSelect(addressTypes, false),
+      addresses: addresses,
+      writeToUsTypeId: writeToUsTypes.length === 1 ? writeToUsTypes[0] : null,
+      errors: errorMsgs.map(errorMsg => { return { text: errorMsg }; }),
+      fatalError: fatalError,
+      primaryPostcodeInvalid: primaryPostcodeInvalid,
+      secondaryPostcodeInvalid: secondaryPostcodeInvalid,
+      updated: updated
+    };
+
+    res.render('courts/tabs/addressesContent', pageData);
+  }
+
+  private validateAddress(address: DisplayAddress, isPrimaryAddress: boolean): AddressValidationResult {
+    const typeErrors = this.validateType(address, isPrimaryAddress);
+    const addressErrors = this.validateAddressLines(address, isPrimaryAddress);
+    const postcodeErrors = this.validatePostcode(address, isPrimaryAddress);
+    const errorPrefix = isPrimaryAddress ? this.primaryAddressPrefix : this.secondaryAddressPrefix;
+
+    return {
+      postcodeValid: postcodeErrors.length === 0,
+      addressValid: addressErrors.length === 0,
+      errors: typeErrors.concat(addressErrors).concat(postcodeErrors).map(error => errorPrefix + error)
+    };
+  }
+
+  private validateType(address: DisplayAddress, isPrimaryAddress: boolean): string[] {
```

Comment:
I got a bit confused with the name for this method (possibly because I am not too bright). I see there is an `validateAddressTypes` method, and a `validateType` method. Is this checking that the address object has a type_id value populated? 

Might be a way to make this more clear and the why behind it, especially in terms of line 140 and why the second condition does not matter if its a primary address